File: js/ui/MiniMap.js
// js/ui/Minimap.js
import { state } from '../state.js';

// A fixed-position, screen-space overview of the ENTIRE world grid —
// drawn after the camera/zoom transform is restored each frame (see
// game.js), so it's unaffected by both. Deliberately shows the whole
// 20x20 grid at once rather than a scrolling/zoomed local view, since
// "each cell in the grid representing a cell in the world" only makes
// sense as a full overview — a player near one edge of the world will
// just show up near one edge of the map, not re-centered.
export class Minimap {
  constructor() {
    // Panel size/position — tune to taste.
    this.size = 240;   // panel is a `size x size` square
    this.margin = 20;  // distance from the canvas' bottom-right corner

    this.backgroundColor = 'rgba(6, 14, 24, 0.72)';
    this.gridColor = 'rgba(80, 200, 255, 0.18)';
    this.borderColor = 'rgba(100, 220, 255, 0.85)';
    this.cornerColor = 'rgba(140, 230, 255, 1)';
    this.labelColor = 'rgba(150, 230, 255, 0.85)';

    this.playerColor = '#ffffff';
    this.playerGlowColor = 'rgba(255,255,255,0.35)';
    this.playerRadius = 5; // deliberately larger than specialRadius, per the brief

    this.specialColor = '#ffd23f';
    this.specialGlowColor = 'rgba(255,210,63,0.5)';
    this.specialRadius = 3;
  }

  // Which planetoids get plotted as the yellow "special" dots. Reads
  // directly from state each call (rather than being passed in) to
  // match how every other system in this codebase already works.
  // Add state.rectPlanet here too if you want the rounded-rect planet
  // included on the map as well.
  getSpecialPlanets() {
    return [state.mazePlanet, state.platformPlanet].filter(Boolean);
  }

  worldToMapPoint(worldX, worldY, x0, y0) {
    return {
      x: x0 + (worldX / state.sceneWidth) * this.size,
      y: y0 + (worldY / state.sceneHeight) * this.size
    };
  }

  drawGlowDot(ctx, x, y, radius, dotColor, glowColor) {
    ctx.beginPath();
    ctx.fillStyle = glowColor;
    ctx.arc(x, y, radius * 2.2, 0, Math.PI * 2);
    ctx.fill();

    ctx.beginPath();
    ctx.fillStyle = dotColor;
    ctx.arc(x, y, radius, 0, Math.PI * 2);
    ctx.fill();
  }

  draw() {
    if (!state.player || !state.sceneWidth || !state.sceneHeight || !state.canvas) return;
    const ctx = state.ctx;
    const gridSize = state.gridSize || 20;

    const x0 = state.canvas.width - this.margin - this.size;
    const y0 = state.canvas.height - this.margin - this.size;

    ctx.save();

    // Panel background
    ctx.fillStyle = this.backgroundColor;
    ctx.fillRect(x0, y0, this.size, this.size);

    // Grid lines — one set of gridSize+1 lines per axis, matching the
    // world's actual cell grid exactly (so each square on the minimap
    // really does correspond to one world cell).
    ctx.strokeStyle = this.gridColor;
    ctx.lineWidth = 1;
    ctx.beginPath();
    for (let i = 0; i <= gridSize; i++) {
      const gx = x0 + (i / gridSize) * this.size;
      ctx.moveTo(gx, y0);
      ctx.lineTo(gx, y0 + this.size);
      const gy = y0 + (i / gridSize) * this.size;
      ctx.moveTo(x0, gy);
      ctx.lineTo(x0 + this.size, gy);
    }
    ctx.stroke();

    // Special planetoids (yellow)
    for (const planet of this.getSpecialPlanets()) {
      const p = this.worldToMapPoint(planet.pos.x, planet.pos.y, x0, y0);
      this.drawGlowDot(ctx, p.x, p.y, this.specialRadius, this.specialColor, this.specialGlowColor);
    }

    // Player (white, larger, drawn last so it's always on top)
    {
      const p = this.worldToMapPoint(state.player.pos.x, state.player.pos.y, x0, y0);
      this.drawGlowDot(ctx, p.x, p.y, this.playerRadius, this.playerColor, this.playerGlowColor);
    }

    // Border + sci-fi corner brackets
    ctx.strokeStyle = this.borderColor;
    ctx.lineWidth = 1.5;
    ctx.strokeRect(x0, y0, this.size, this.size);

    const bracket = Math.min(20, this.size * 0.12);
    ctx.strokeStyle = this.cornerColor;
    ctx.lineWidth = 2;
    const corners = [
      [x0, y0, 1, 1],
      [x0 + this.size, y0, -1, 1],
      [x0, y0 + this.size, 1, -1],
      [x0 + this.size, y0 + this.size, -1, -1]
    ];
    for (const [cx, cy, dx, dy] of corners) {
      ctx.beginPath();
      ctx.moveTo(cx, cy + bracket * dy);
      ctx.lineTo(cx, cy);
      ctx.lineTo(cx + bracket * dx, cy);
      ctx.stroke();
    }

    // Label
    ctx.fillStyle = this.labelColor;
    ctx.font = '12px monospace';
    ctx.textAlign = 'left';
    ctx.fillText('SECTOR MAP', x0 + 8, y0 - 8);

    ctx.restore();
  }
}

File: lua/ui/MiniMap.lua
-- lua/ui/Minimap.lua
--
-- Screen-space overview of the full world grid.
-- Belt marker is a static thick circular annulus around the sun,
-- not a box and not live planetoid positions.

local state = require("lua.state")

local Minimap = {}
Minimap.__index = Minimap

function Minimap.new()
  local self = setmetatable({}, Minimap)

  self.size = 240
  self.sizeFraction = 0.20
  self.margin = 20

  self.backgroundColor = { 6/255, 14/255, 24/255, 0.72 }
  self.gridColor       = { 80/255, 200/255, 255/255, 0.18 }
  self.borderColor     = { 100/255, 220/255, 255/255, 0.85 }
  self.cornerColor     = { 140/255, 230/255, 255/255, 1 }
  self.labelColor      = { 150/255, 230/255, 255/255, 0.85 }

  self.playerColor     = { 1, 1, 1, 1 }
  self.playerGlowColor = { 1, 1, 1, 0.35 }
  self.playerRadius    = 5

  self.specialColor     = { 1, 210/255, 63/255, 1 }
  self.specialGlowColor = { 1, 210/255, 63/255, 0.5 }
  self.specialRadius    = 3

  self.sunColor     = { 1, 0.75, 0.2, 1 }
  self.sunGlowColor = { 1, 0.55, 0.1, 0.45 }

  self.beltDotColor = { 1, 77/255, 77/255, 0.78 }
  self.beltDotCount = 220
  self._beltDots = nil

  return self
end

function Minimap:refreshSize()
  local w = love.graphics.getWidth()
  self.size = math.max(120, math.floor(w * (self.sizeFraction or 0.25)))
end

function Minimap:getSpecialPlanets()
  local list = {}
  if state.mazePlanet then table.insert(list, state.mazePlanet) end
  if state.platformPlanet then table.insert(list, state.platformPlanet) end
  if state.skyDomePlanet then table.insert(list, state.skyDomePlanet) end
  return list
end

function Minimap:worldToMapPoint(worldX, worldY, x0, y0)
  local sw = state.sceneWidth or 1
  local sh = state.sceneHeight or 1
  return {
    x = x0 + (worldX / sw) * self.size,
    y = y0 + (worldY / sh) * self.size,
  }
end

function Minimap:drawGlowDot(x, y, radius, color, glowColor)
  love.graphics.setColor(glowColor)
  love.graphics.circle("fill", x, y, radius * 2.2)
  love.graphics.setColor(color)
  love.graphics.circle("fill", x, y, radius)
end

function Minimap:ensureBeltDots()
  if self._beltDots then return end

  local worldGen = require("lua.setup.worldGen")
  local cellSize = worldGen.CELL_SIZE or 3000
  local rMin = (worldGen.BELT_INNER_CELLS or 2) * cellSize
  local rMax = (worldGen.BELT_OUTER_CELLS or 3) * cellSize
  local sunX = (state.sceneWidth or cellSize * 20) / 2
  local sunY = (state.sceneHeight or cellSize * 20) / 2

  local dots = {}
  local rng = love.math.newRandomGenerator(20260821)
  local n = self.beltDotCount or 220

  for _ = 1, n do
    local ang = rng:random() * math.pi * 2
    -- Uniform in annulus area: r = sqrt(u * (R2^2 - R1^2) + R1^2)
    local u = rng:random()
    local r = math.sqrt(u * (rMax * rMax - rMin * rMin) + rMin * rMin)
    table.insert(dots, {
      x = sunX + math.cos(ang) * r,
      y = sunY + math.sin(ang) * r,
    })
  end

  self._beltDots = dots
end

function Minimap:draw()
  self:refreshSize()
  self:ensureBeltDots()

  local winW = love.graphics.getWidth()
  local winH = love.graphics.getHeight()
  local x0 = winW - self.size - self.margin
  local y0 = winH - self.size - self.margin

  love.graphics.setColor(self.backgroundColor)
  love.graphics.rectangle("fill", x0, y0, self.size, self.size)

  local gridSize = (state.gridSize or 20)
  local cellPx = self.size / gridSize
  love.graphics.setColor(self.gridColor)
  love.graphics.setLineWidth(1)
  for i = 1, gridSize - 1 do
    local t = x0 + i * cellPx
    love.graphics.line(t, y0, t, y0 + self.size)
    local u = y0 + i * cellPx
    love.graphics.line(x0, u, x0 + self.size, u)
  end

  love.graphics.setColor(self.beltDotColor)
  for _, d in ipairs(self._beltDots) do
    local m = self:worldToMapPoint(d.x, d.y, x0, y0)
    love.graphics.circle("fill", m.x, m.y, 1.15)
  end

  if state.sun and state.sun.pos then
    local sunMap = self:worldToMapPoint(state.sun.pos.x, state.sun.pos.y, x0, y0)
    local sunR = 6
    if state.sun.radius and state.sceneWidth then
      sunR = math.max(4, (state.sun.radius / state.sceneWidth) * self.size)
      sunR = math.min(sunR, self.size * 0.08)
    end
    self:drawGlowDot(sunMap.x, sunMap.y, sunR, self.sunColor, self.sunGlowColor)
  end

  for _, planet in ipairs(self:getSpecialPlanets()) do
    if planet and planet.pos then
      local p = self:worldToMapPoint(planet.pos.x, planet.pos.y, x0, y0)
      self:drawGlowDot(p.x, p.y, self.specialRadius, self.specialColor, self.specialGlowColor)
    end
  end

  if state.player and state.player.pos then
    local playerMap = self:worldToMapPoint(state.player.pos.x, state.player.pos.y, x0, y0)
    self:drawGlowDot(playerMap.x, playerMap.y, self.playerRadius, self.playerColor, self.playerGlowColor)
  end

  love.graphics.setColor(self.borderColor)
  love.graphics.setLineWidth(1.5)
  love.graphics.rectangle("line", x0, y0, self.size, self.size)

  local bracket = math.min(20, self.size * 0.12)
  love.graphics.setColor(self.cornerColor)
  love.graphics.setLineWidth(2)

  local corners = {
    { x0, y0, 1, 1 },
    { x0 + self.size, y0, -1, 1 },
    { x0, y0 + self.size, 1, -1 },
    { x0 + self.size, y0 + self.size, -1, -1 },
  }
  for _, c in ipairs(corners) do
    local cx, cy, dx, dy = c[1], c[2], c[3], c[4]
    love.graphics.line(cx, cy + bracket * dy, cx, cy)
    love.graphics.line(cx, cy, cx + bracket * dx, cy)
  end

  love.graphics.setColor(self.labelColor)
  love.graphics.print("SECTOR MAP", x0 + 8, y0 - 18)

  love.graphics.setColor(1, 1, 1, 1)
  love.graphics.setLineWidth(1)
end

return Minimap